### PR TITLE
fix: errored stream is disturbed

### PR DESCRIPTION
**Pull request by ronag**

### CODE REVIEW

File: test/node-fetch/main.js [hunk 1]
```diff
@@ -882,7 +882,7 @@ describe('node-fetch', () => {
           return expect(res.text())
             .to.eventually.be.rejected
             .and.be.an.instanceof(Error)
-            .and.have.property('name', 'AbortError')
+            .and.have.property('name', 'TypeError')
```
Comment (ronag) [inline test/node-fetch/main.js:885]: @dominic I'm actually unsure what the spec expects here. Would you mind help clarifying?

**Comment by ronag** [comment]
@szmarczak This is a spec thing. Care to help clarifying what is the correct behaviour.

**Comment by ronag** [comment]
Screenshot 2021-12-08 at 21 03 30

**Comment by ronag** [comment]
@jimmywarting: Any thoughts? I believe node-fetch fails with an `AbortError` in this test.

**Comment by jimmywarting** [comment]
Umm. is there any wpt test or something in the spec that mention what should actually be happening?

File: lib/fetch/index.js [hunk 2]
```diff
@@ -1742,8 +1742,9 @@ function httpNetworkFetch (
             controller.enqueue(new Uint8Array(bytes))
 
             // 8. If stream is errored, then terminate the ongoing fetch.
-            if (stream[kError]) {
-              this.context.terminate({ reason: stream[kError] })
+            // TODO (perf): This check is slow,
+            if (/state: 'errored'/.test(nodeUtil.inspect(stream))) {
+              this.context.terminate()
```
Comment (mcollina) [inline lib/fetch/index.js:1747]: Maybe create an issue for this?

Comment (ronag) [reply lib/fetch/index.js:1747]: Where? I think I already created one for @jasnell in node core. But I don't think this is something we can fix unless undici is in core.

Comment (ronag) [reply lib/fetch/index.js:1747]: https://github.com/nodejs/node/pull/41121

**Comment by szmarczak** [comment]
Sorry for my inactivity, had a few tests and an exam in uni. Will take a look now.

File: lib/core/util.js [hunk 3]
```diff
@@ -244,13 +244,32 @@ function validateHandler (handler, method, upgrade) {
...
+
   const state = body && body._readableState
-  return !!(body && (
-    (stream.isDisturbed && stream.isDisturbed(body)) ||
+  return !!(
     body[kBodyUsed] ||
     body.readableDidRead || (state && state.dataEmitted) ||
-    isAborted(body)
-  ))
+    isReadableAborted(body)
```
Comment (szmarczak) [inline lib/core/util.js:259]: This could be `isDestroyed`, no? Is there a reason why we check `endEmitted`?

File: test/node-fetch/main.js (hunk 1 above)
Comment (szmarczak) [reply test/node-fetch/main.js:885]: It should be `AbortError`.

https://fetch.spec.whatwg.org/#fetch-method -> https://fetch.spec.whatwg.org/#abort-fetch

**Comment by szmarczak** [comment]
@jimmywarting https://github.com/whatwg/streams/issues/1025 tl;dr unfortunately it's up to Node.js, because they made the mistake to make fetch access stream internals.

**Review by mcollina — COMMENTED** [review]
Can you please add a unit test?

File: lib/core/util.js (hunk 3 above)
Comment (ronag) [reply lib/core/util.js:259]: Yes. If it is destroyed before emitting end then and only the is it aborted.

**Comment by ronag** [comment]
> Can you please add a unit test?

There is already unit test?

File: test/node-fetch/main.js (hunk 1 above)
Comment (ronag) [reply test/node-fetch/main.js:885]: > If response’s body is not null and is readable, then error response’s body with error.

This does not say how that error should propagate through e.g. `.text()`

Comment (ronag) [reply test/node-fetch/main.js:885]: erroring a stream makes it disturbed and the reading algorithm says to throw a TypeError if disturbed.

Comment (ronag) [reply test/node-fetch/main.js:885]: @szmarczak I think this is the relevant part: https://fetch.spec.whatwg.org/#concept-body-consume-body

Comment (ronag) [reply test/node-fetch/main.js:885]: Remember it's not fetch that throws here, we already have resolved the promise, it's the body mixing methods that throw here. wdyt?

**Comment by mcollina** [comment]
I missed it. CI is failing.

**Review by mcollina — APPROVED** [review]
lgtm

File: test/node-fetch/main.js (hunk 1 above)
Comment (szmarczak) [reply test/node-fetch/main.js:885]: Makes sense, I guess you're right.

```js
const start = Date.now();
const controller = new AbortController();
const response = await fetch('https://speed.hetzner.de/100MB.bin', {signal: controller.signal});
console.log(Date.now() - start, response.bodyUsed);
(async () => {
    try {
        await response.text();
        console.log(Date.now() - start, 'read');
    } catch (error) {
        console.error(Date.now() - start, error);
    }
})();
console.log(Date.now() - start, 'aborting');
controller.abort();
```

Google Chrome: `DOMException: The user aborted a request.`
Firefox: `DOMException: The operation was aborted.`

**Review by szmarczak — APPROVED** [review]
Indeed, it should be `TypeError` 👍🏼

**Pull request merged**